### PR TITLE
ENH: Eigen3: Use always find_package for Eigen3

### PR DESCRIPTION
**Pull request by phcerdan**
Adds the file `itkExternal_Eigen3.cmake` to manage the Eigen3 configuration.

Removes the usage of `add_subdirectory` to include the internal Eigen.
Using instead `execute_process` to generate internal Eigen Target and Config.cmake files
at configure time, so `find_package` can be used afterwards.

The problem with `add_subdirectory` is that Remote module using internal libraries
with `find_package(Eigen3)`, are incompatible with the Eigen target generated internally
(because it already exists).
This is not a problem when the remote module is built
out-of-source (with `find_package(ITK)`), but it fails when the module is enabled in-source (using `Module_X=ON`).

This is a clean solution, providing third-parties interested in using Eigen to re-use
the Eigen shipped internally in ITK.
But without the need of a different notation, like the `#include "itkeigen/Eigen/X"` used internally in ITK,
but directly `#include "Eigen/X"`.
The idea is to help users to integrate with existing projects using Eigen in a effortless way.
To re-use the ITK internal Eigen, users just need:
```cmake
option(use_internal_itk_eigen "Use the Eigen3 provided by ITK" OFF)
find_package(ITK)
if(use_internal_itk_eigen AND NOT ITK_USE_SYSTEM_EIGEN)
   set(Eigen3_DIR ${ITKInternalEigen3_DIR})
endif()
find_package(Eigen3)
```

## PR Checklist

- [X] :no_entry_sign: [Makes breaking changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes)
- [X] :no_entry_sign: [Makes design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes)
- [X] :no_entry_sign: Adds the License notice to new files.
- [X] :no_entry_sign: Adds Python wrapping.
- [X] :no_entry_sign: Adds tests and baseline comparison (quantitative).
- [X] :no_entry_sign: [Adds test data](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/UploadBinaryData.md).
- [X] :no_entry_sign: Adds Examples to [ITKExamples](https://github.com/InsightSoftwareConsortium/ITKExamples)
- [X] :no_entry_sign: Adds Documentation.

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by phcerdan** [comment]
Also updates Eigen3 to the top of the branch 3.3

**Comment by phcerdan** [comment]
> I would expect more changes given the huge shortlog. Are you sure you selected the right commits for shortlog?

Yeah, the shortlog is huge, but the diff files are not that much at all...

I started from scratch, pushing ITK changes (2-3 commits) on top of the upstream 3.3.
Not sure about the internal of UpdateFromStream script, but it seems the shortlog is getting basically all the Eigen commits from the start of time.
I will have a look to only include those that really differ from the current version.

File: CMake/itkExternal_Eigen3.cmake [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# to avoid introducing GPL code from Eigen3 internally in ITK.
+option(ITK_USE_EIGEN_MPL2_ONLY "Set compile definition EIGEN_MPL2_ONLY for ITKInternalEigen3." OFF)
+mark_as_advanced(ITK_USE_EIGEN_MPL2_ONLY)
+if(ITK_USE_EIGEN_MPL2_ONLY)
+  list(APPEND _additional_external_project_args "-DITK_USE_EIGEN_MPL2_ONLY:BOOL=ON")
+endif()
+
+if(ITK_USE_SYSTEM_EIGEN)
+  find_package( Eigen )
+  # TODO do we need include_directories?
```
[the remark this answers is not in the document]
Comment (phcerdan) [reply CMake/itkExternal_Eigen3.cmake]: Thanks, will remove.

**Comment by phcerdan** [comment]
The problem of the huge shortlog is some kind of rebase upstream that has rewritten history, same commits have different SHA.
This upgrade has only 9 new commits from upstream.
I can manually remove all the incorrect entries from the shortlog. Incoming.

**Comment by phcerdan** [comment]
I think shortlog issue is because now upstream is officially on gitlab (they moved from bitbucket). Before we were using a git mirror created by some user.

They way UpdateFromUpstream works, not sure if we can avoid the long shortlog. I am all ears.

Updated to handle your reviews. Also moved the _eigen3_build_dir to ITK_BINARY_DIR/ITKInternalEigen3.
Not sure if you would prefer another location for the "build" (nothing to compile really, but still needed.

**Comment by phcerdan** [comment]
There is a warning in python test `PythonGetNameOfClass`, but seems unrelated.

**Comment by thewtex** [comment]
> There is a warning in python test PythonGetNameOfClass, but seems unrelated.

@phcerdan could you please rebase and push so we can verify?

**Comment by phcerdan** [comment]
To avoid the shortlog mess, I have recreated the itkeigen subtree, so UpdateThirdPartyFromUpstream treats it as a new tree, avoiding pulling all the commits from upstream.

I had to remove the itkeigen folder before doing this to avoid subtree conflicts..., that's why there is an extra commit.

I think is cleaner now, despite the extra commit.

**Comment by dzenanz** [comment]
@bradking ghostflow complains about a root commit. How to override that? Merge using git instead of GitHub? Or something else?

**Comment by thewtex** [comment]
>  I have recreated the itkeigen subtree, 

Ah, sorry about that -- I forgot that this had a subtree. Thanks for the extra effort!

**Comment by bradking** [comment]
@dzenanz an admin user needs to do the merge and push by hand.  I've done the merge.

**Comment by phcerdan** [comment]
Thanks @bradking.

All CI checks, including python, are green

**Comment by hjmjohnson** [comment]
@phcerdan "Total Variation" Remote module does not build, and I think it is related to these changes.

https://github.com/InsightSoftwareConsortium/ITKTotalVariation/blob/master/CMakeLists.txt#L30

Could you please have a look?

**Comment by phcerdan** [comment]
Thanks @hjmjohnson, PR incoming there.

**Pull request merged**